// date-picker-directive/esm5/lib/date-picker.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Directive, EventEmitter, Input, NgZone, Output, HostListener, Renderer2 } from "@angular/core";
import { ComponentFactoryResolver, ViewContainerRef } from '@angular/core';
import { DatePickerDirectiveComponent } from './date-picker-directive.component';
var DatePickerDirective = /** @class */ (function () {
    function DatePickerDirective(ngZone, renderer, viewContainer, componentFactoryResolver) {
        this.ngZone = ngZone;
        this.renderer = renderer;
        this.viewContainer = viewContainer;
        this.componentFactoryResolver = componentFactoryResolver;
        this.dateselected = new EventEmitter();
        this.ngModelChanged = new EventEmitter();
        this.componentRef = null;
        this.pickerLoaded = false;
    }
    /**
     * @param {?} event
     * @return {?}
     */
    DatePickerDirective.prototype.onFocus = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        var _this = this;
        this.renderer.setStyle(this.viewContainer.element.nativeElement.parentElement, 'position', 'relative');
        var /** @type {?} */ height = this.viewContainer.element.nativeElement.offsetHeight;
        this.viewContainer.clear();
        var /** @type {?} */ componentFactory = this.componentFactoryResolver.resolveComponentFactory(DatePickerDirectiveComponent);
        this.componentRef = this.viewContainer.createComponent(componentFactory);
        this.componentRef.instance.minDate = this.minDate;
        this.componentRef.instance.maxDate = this.maxDate;
        this.componentRef.instance.singleDatePicker = this.singleDatePicker;
        this.componentRef.instance.startDate = this.startDate;
        this.componentRef.instance.endDate = this.endDate;
        var /** @type {?} */ elm = /** @type {?} */ (document.querySelector('.rl-drpicker'));
        elm.setAttribute('style', 'top:' + parseInt(height + 10) + 'px');
        var /** @type {?} */ rect = elm.getBoundingClientRect();
        var /** @type {?} */ l = rect.left;
        var /** @type {?} */ w = rect.width;
        var /** @type {?} */ docH = document.getElementsByTagName("body")[0].offsetHeight;
        var /** @type {?} */ docW = document.getElementsByTagName("body")[0].offsetWidth;
        var /** @type {?} */ isEntirelyVisible = (l + w <= docW);
        if (!isEntirelyVisible) {
            elm.classList.add('edge');
        }
        else {
            elm.classList.remove('edge');
        }
        this.componentRef.instance.close.subscribe(function () {
            _this.componentRef.destroy();
        });
        this.componentRef.instance.dateSelected.subscribe(function (val) {
            _this.dateselected.emit(val);
            _this.componentRef.destroy();
            _this.pickerLoaded = false;
        });
    };
    /**
     * @param {?} targetElement
     * @return {?}
     */
    DatePickerDirective.prototype.onClick = /**
     * @param {?} targetElement
     * @return {?}
     */
    function (targetElement) {
        if (this.componentRef) {
            var /** @type {?} */ clickedDp = this.componentRef.location.nativeElement.contains(targetElement);
            var /** @type {?} */ clickedInput = this.viewContainer.element.nativeElement == targetElement;
            var /** @type {?} */ inDp = (targetElement.parentElement ? targetElement.parentElement.classList.contains('calendar-row') : false);
            if (!clickedDp && !clickedInput && !inDp) {
                if (this.componentRef) {
                    this.componentRef.destroy();
                }
            }
        }
    };
    DatePickerDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[date-picker]'
                },] },
    ];
    /** @nocollapse */
    DatePickerDirective.ctorParameters = function () { return [
        { type: NgZone },
        { type: Renderer2 },
        { type: ViewContainerRef },
        { type: ComponentFactoryResolver }
    ]; };
    DatePickerDirective.propDecorators = {
        minDate: [{ type: Input }],
        maxDate: [{ type: Input }],
        singleDatePicker: [{ type: Input }],
        linkedCalendars: [{ type: Input }],
        startDate: [{ type: Input }],
        ngModel: [{ type: Input }],
        endDate: [{ type: Input }],
        dateselected: [{ type: Output }],
        ngModelChanged: [{ type: Output }],
        onFocus: [{ type: HostListener, args: ['focus', ['$event'],] }],
        onClick: [{ type: HostListener, args: ['document:click', ['$event.target'],] }]
    };
    return DatePickerDirective;
}());
export { DatePickerDirective };
function DatePickerDirective_tsickle_Closure_declarations() {
    /** @type {?} */
    DatePickerDirective.prototype.minDate;
    /** @type {?} */
    DatePickerDirective.prototype.maxDate;
    /** @type {?} */
    DatePickerDirective.prototype.singleDatePicker;
    /** @type {?} */
    DatePickerDirective.prototype.linkedCalendars;
    /** @type {?} */
    DatePickerDirective.prototype.startDate;
    /** @type {?} */
    DatePickerDirective.prototype.ngModel;
    /** @type {?} */
    DatePickerDirective.prototype.endDate;
    /** @type {?} */
    DatePickerDirective.prototype.dateselected;
    /** @type {?} */
    DatePickerDirective.prototype.ngModelChanged;
    /** @type {?} */
    DatePickerDirective.prototype.componentRef;
    /** @type {?} */
    DatePickerDirective.prototype.pickerLoaded;
    /** @type {?} */
    DatePickerDirective.prototype.ngZone;
    /** @type {?} */
    DatePickerDirective.prototype.renderer;
    /** @type {?} */
    DatePickerDirective.prototype.viewContainer;
    /** @type {?} */
    DatePickerDirective.prototype.componentFactoryResolver;
}
